File: C++/Laba3sem2v2/binaryTree.cpp
#include "binaryTree.h"
#include <cstring>
#include <fstream>
#include <iostream>
#include <queue>
#include <string>

using namespace std;

BinaryTree::BinaryTree() : root(nullptr) {}

BinaryTree::~BinaryTree() {
    if (root) {
        deleteTree(root);
        root = nullptr;
    }
}

void BinaryTree::deleteTree(Node* node) {
    if (!node) return;
    // Сначала удаляем поддеревья, потом сам узел
    deleteTree(node->left);
    deleteTree(node->right);
    delete node;
}


void BinaryTree::insert(const std::string& key) {
    root = insert(root, key);
}

BinaryTree::Node* BinaryTree::insert(Node* node, const std::string& key) {
    if (key.empty())
        return node;

    if (node == nullptr)
        return new Node(key);

    if (key < node->key)
        node->left = insert(node->left, key);
    else if (key > node->key)
        node->right = insert(node->right, key);

    return node;
}

bool BinaryTree::search(const std::string& key) {
    Node* current = root;

    while (current != nullptr) {
        if (key == current->key)
            return true;
        else if (key < current->key)
            current = current->left;
        else
            current = current->right;
    }

    return false;
}

// ИСПРАВЛЕНО: бинарная запись с префиксным обходом
void BinaryTree::saveNode(Node* n, std::ofstream& f) {
    bool exists = (n != nullptr);
    f.write((char*)&exists, sizeof(exists));
    if (!exists) return;

    size_t len = n->key.size();
    f.write((char*)&len, sizeof(len));
    f.write(n->key.c_str(), len);

    saveNode(n->left, f);
    saveNode(n->right, f);
}

// ИСПРАВЛЕНО: бинарное чтение с префиксным обходом
void BinaryTree::loadNode(std::ifstream& f) {
    // Очищаем текущее дерево
    deleteTree(root);
    root = nullptr;

    // Рекурсивно загружаем узлы
    root = loadNodeRecursive(f);
}

// Добавляем новый вспомогательный метод для рекурсивной загрузки
BinaryTree::Node* BinaryTree::loadNodeRecursive(std::ifstream& f) {
    bool exists;
    if (!f.read((char*)&exists, sizeof(exists))) {
        return nullptr;
    }

    if (!exists) return nullptr;

    size_t len;
    f.read((char*)&len, sizeof(len));
    if (f.fail()) return nullptr;

    // Проверяем, что len разумная (не слишком большая)
    if (len > 10000) { // Защита от некорректных данных
        return nullptr;
    }

    string key(len, '\0');
    f.read(&key[0], len);
    if (f.fail()) return nullptr;

    Node* node = nullptr;
    try {
        node = new Node(key);
        node->left = loadNodeRecursive(f);
        node->right = loadNodeRecursive(f);
    }
    catch (const std::bad_alloc&) {
        // Если не хватает памяти, возвращаем nullptr
        return nullptr;
    }

    return node;
}

void BinaryTree::saveToFile(const string& fileName) {
    if (fileName.empty()) return; // ДОБАВЛЕНА ПРОВЕРКА
    ofstream file(fileName, ios::binary);  // ВАЖНО: бинарный режим
    if (!file.is_open()) {
        cout << "Не удалось открыть файл для записи." << endl;
        return;
    }
    saveNode(root, file);
    file.close();
}

void BinaryTree::loadFromFile(const string& fileName) {
    if (fileName.empty()) return;

    // Очищаем текущее дерево
    if (root) {
        deleteTree(root);
        root = nullptr;
    }

    ifstream file(fileName, ios::binary);
    if (!file.is_open()) {
        return;
    }

    root = loadNodeRecursive(file);
    file.close();
}

bool BinaryTree::isFullNode(Node* node) const {
    if (!node) return true;

    bool hasLeft = (node->left != nullptr);
    bool hasRight = (node->right != nullptr);

    // Если узел имеет ровно двух потомков
    if (hasLeft && hasRight) {
        return isFullNode(node->left) && isFullNode(node->right);
    }
    // Если узел не имеет потомков (лист)
    else if (!hasLeft && !hasRight) {
        return true;
    }
    // Если узел имеет ровно одного потомка
    else {
        return false;
    }
}

bool BinaryTree::isFull() const {
    if (!root) return true; // Пустое дерево считаем полным
    return isFullNode(root);
}


void BinaryTree::inorder(Node* node) {
    if (!node) return;
    inorder(node->left);
    cout << node->key << " ";
    inorder(node->right);
}

void BinaryTree::preorder(Node* node) {
    if (!node) return;
    cout << node->key << " ";
    preorder(node->left);
    preorder(node->right);
}

void BinaryTree::postorder(Node* node) {
    if (!node) return;
    postorder(node->left);
    postorder(node->right);
    cout << node->key << " ";
}

void BinaryTree::printInorder() {
    cout << "Inorder traversal: ";
    inorder(root);
    cout << endl;
}

void BinaryTree::printPreorder() {
    cout << "Preorder traversal: ";
    preorder(root);
    cout << endl;
}

void BinaryTree::printPostorder() {
    cout << "Postorder traversal: ";
    postorder(root);
    cout << endl;
}

void BinaryTree::printBFS() {
    if (!root) {
        cout << endl;
        return;
    }

    queue<Node*> q;
    q.push(root);
    bool first = true;

    while (!q.empty()) {
        Node* cur = q.front();
        q.pop();

        if (!first) cout << " ";
        cout << cur->key;
        first = false;

        if (cur->left) q.push(cur->left);
        if (cur->right) q.push(cur->right);
    }
    cout << endl;
}

void runBinaryTree(int argc, char* argv[]) {
    BinaryTree tree;

    string fileName;
    string query;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--file") == 0 && i + 1 < argc) {
            fileName = argv[i + 1];
            i++;
        }
        else if (strcmp(argv[i], "--query") == 0 && i + 1 < argc) {
            query = argv[i + 1];
            i++;
        }
    }

    // Загружаем дерево только если файл существует и не пустой
    if (!fileName.empty()) {
        // Проверяем, существует ли файл
        ifstream test(fileName);
        if (test.good()) {
            test.close();
            tree.loadFromFile(fileName);
        }
    }

    string command;
    string arg;
    size_t pos = query.find(' ');
    if (pos != string::npos) {
        command = query.substr(0, pos);
        arg = query.substr(pos + 1);
    }
    else {
        command = query;
    }

    if (command == "TINSERT") {
        if (!arg.empty()) {
            tree.insert(arg);
            if (!fileName.empty()) tree.saveToFile(fileName);
        }
    }
    else if (command == "TGET") {
        if (tree.search(arg)) cout << arg << endl;
        else cout << "NOT_FOUND" << endl;
    }
    else if (command == "TFULL") {
        cout << (tree.isFull() ? "true" : "false") << endl;
    }
    else if (command == "TSEARCH") {
        cout << (tree.search(arg) ? "true" : "false") << endl;
    }
    else if (command == "TINORDER") {
        tree.printInorder();
    }
    else if (command == "TPREORDER") {
        tree.printPreorder();
    }
    else if (command == "TPOSTORDER") {
        tree.printPostorder();
    }
    else if (command == "TBFS") {
        cout << "BFS обход: ";
        tree.printBFS();
    }
}

File: sem3l3go/binarytree/binarytree.go
package binarytree

import (
	"encoding/binary"
	"fmt"
	"os"
	"strings"
)

// Node представляет узел бинарного дерева
type Node struct {
	Key   string
	Left  *Node
	Right *Node
}

// BinaryTree представляет бинарное дерево поиска
type BinaryTree struct {
	Root *Node
}

// NewBinaryTree создает новое дерево
func NewBinaryTree() *BinaryTree {
	return &BinaryTree{
		Root: nil,
	}
}

// Insert вставляет ключ в дерево
func (bt *BinaryTree) Insert(key string) {
	bt.Root = bt.insertNode(bt.Root, key)
}

func (bt *BinaryTree) insertNode(node *Node, key string) *Node {
	if key == "" {
		return node
	}
	if node == nil {
		return &Node{Key: key}
	}
	if key < node.Key {
		node.Left = bt.insertNode(node.Left, key)
	} else if key > node.Key {
		node.Right = bt.insertNode(node.Right, key)
	}
	return node
}

// Search ищет ключ в дереве
func (bt *BinaryTree) Search(key string) bool {
	return bt.searchNode(bt.Root, key)
}

func (bt *BinaryTree) searchNode(node *Node, key string) bool {
	if node == nil {
		return false
	}
	if key == node.Key {
		return true
	}
	if key < node.Key {
		return bt.searchNode(node.Left, key)
	}
	return bt.searchNode(node.Right, key)
}

// IsFull проверяет, является ли дерево полным
// IsFull проверяет, является ли дерево полным
func (bt *BinaryTree) IsFull() bool {
	if bt.Root == nil {
		return true
	}
	return bt.isFullNode(bt.Root)
}

func (bt *BinaryTree) isFullNode(node *Node) bool {
	if node == nil {
		return true
	}

	// Если узел - лист (нет детей)
	if node.Left == nil && node.Right == nil {
		return true
	}

	// Если узел имеет обоих детей
	if node.Left != nil && node.Right != nil {
		return bt.isFullNode(node.Left) && bt.isFullNode(node.Right)
	}

	// Если узел имеет только одного ребенка
	return false
}

// PrintInorder выводит in-order обход
func (bt *BinaryTree) PrintInorder() {
	fmt.Print("Inorder traversal: ")
	bt.inorder(bt.Root)
	fmt.Println()
}

func (bt *BinaryTree) inorder(node *Node) {
	if node == nil {
		return
	}
	bt.inorder(node.Left)
	fmt.Printf("%s ", node.Key)
	bt.inorder(node.Right)
}

// PrintPreorder выводит pre-order обход
func (bt *BinaryTree) PrintPreorder() {
	fmt.Print("Preorder traversal: ")
	bt.preorder(bt.Root)
	fmt.Println()
}

func (bt *BinaryTree) preorder(node *Node) {
	if node == nil {
		return
	}
	fmt.Printf("%s ", node.Key)
	bt.preorder(node.Left)
	bt.preorder(node.Right)
}

// PrintPostorder выводит post-order обход
func (bt *BinaryTree) PrintPostorder() {
	fmt.Print("Postorder traversal: ")
	bt.postorder(bt.Root)
	fmt.Println()
}

func (bt *BinaryTree) postorder(node *Node) {
	if node == nil {
		return
	}
	bt.postorder(node.Left)
	bt.postorder(node.Right)
	fmt.Printf("%s ", node.Key)
}

// PrintBFS выводит BFS обход
func (bt *BinaryTree) PrintBFS() {
	if bt.Root == nil {
		fmt.Println()
		return
	}

	queue := []*Node{bt.Root}
	first := true

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		if !first {
			fmt.Print(" ")
		}
		fmt.Print(current.Key)
		first = false

		if current.Left != nil {
			queue = append(queue, current.Left)
		}
		if current.Right != nil {
			queue = append(queue, current.Right)
		}
	}
	fmt.Println()
}

// saveNode рекурсивно сохраняет узел в бинарном формате
func (bt *BinaryTree) saveNode(node *Node, file *os.File) error {
	exists := node != nil
	if err := binary.Write(file, binary.LittleEndian, exists); err != nil {
		return err
	}
	if !exists {
		return nil
	}

	if err := binary.Write(file, binary.LittleEndian, int32(len(node.Key))); err != nil {
		return err
	}
	if _, err := file.Write([]byte(node.Key)); err != nil {
		return err
	}

	if err := bt.saveNode(node.Left, file); err != nil {
		return err
	}
	return bt.saveNode(node.Right, file)
}

// loadNode рекурсивно загружает узел из бинарного файла
func (bt *BinaryTree) loadNode(file *os.File) (*Node, error) {
	var exists bool
	if err := binary.Read(file, binary.LittleEndian, &exists); err != nil {
		return nil, err
	}
	if !exists {
		return nil, nil
	}

	var keyLen int32
	if err := binary.Read(file, binary.LittleEndian, &keyLen); err != nil {
		return nil, err
	}

	keyBytes := make([]byte, keyLen)
	if _, err := file.Read(keyBytes); err != nil {
		return nil, err
	}

	node := &Node{Key: string(keyBytes)}

	left, err := bt.loadNode(file)
	if err != nil {
		return nil, err
	}
	node.Left = left

	right, err := bt.loadNode(file)
	if err != nil {
		return nil, err
	}
	node.Right = right

	return node, nil
}

// SaveToFile сохраняет дерево в бинарный файл
func (bt *BinaryTree) SaveToFile(filename string) error {
	if filename == "" {
		return nil
	}
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	return bt.saveNode(bt.Root, file)
}

// LoadFromFile загружает дерево из бинарного файла
func (bt *BinaryTree) LoadFromFile(filename string) error {
	if filename == "" {
		return nil
	}
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	root, err := bt.loadNode(file)
	if err != nil {
		return err
	}
	bt.Root = root
	return nil
}

// RunBinaryTree выполняет команды над деревом
func RunBinaryTree(args []string) {
	tree := NewBinaryTree()
	var filename string
	var query string

	for i := 0; i < len(args); i++ {
		if args[i] == "--file" && i+1 < len(args) {
			filename = args[i+1]
			i++
		} else if args[i] == "--query" && i+1 < len(args) {
			query = args[i+1]
			i++
		}
	}

	if filename != "" {
		tree.LoadFromFile(filename)
	}

	parts := strings.SplitN(query, " ", 2)
	command := parts[0]
	arg := ""
	if len(parts) > 1 {
		arg = parts[1]
	}

	switch command {
	case "TINSERT":
		if arg != "" {
			tree.Insert(arg)
			if filename != "" {
				tree.SaveToFile(filename)
			}
		}
	case "TGET", "TSEARCH":
		if tree.Search(arg) {
			if command == "TGET" {
				fmt.Println(arg)
			} else {
				fmt.Println("true")
			}
		} else {
			if command == "TGET" {
				fmt.Println("NOT_FOUND")
			} else {
				fmt.Println("false")
			}
		}
	case "TFULL":
		if tree.IsFull() {
			fmt.Println("true")
		} else {
			fmt.Println("false")
		}
	case "TINORDER":
		tree.PrintInorder()
	case "TPREORDER":
		tree.PrintPreorder()
	case "TPOSTORDER":
		tree.PrintPostorder()
	case "TBFS":
		fmt.Print("BFS обход: ")
		tree.PrintBFS()
	}
}
